atomics/include/desul/atomics/Generic.hpp

```diff
@@ -136,169 +136,88 @@ DESUL_INLINE_FUNCTION bool atomic_is_lock_free() noexcept {
-                  MemoryScope scope) {
-  union U {
-    int64_t i;
-    T t;
-    DESUL_INLINE_FUNCTION U() {}
-  } oldval, assume, newval;
+template<std::size_t N>
+struct atomic_compare_exchange_type;
 
-  oldval.t = *dest;
-
-  do {
-    assume.i = oldval.i;
-    newval.t = op.apply(assume.t, val);
-    oldval.i = desul::atomic_compare_exchange(
-        (int64_t*)dest, assume.i, newval.i, order, scope);
-  } while (assume.i != oldval.i);
-
-  return oldval.t;
-}
-
-template <class Oper, typename T, class MemoryOrder, class MemoryScope>
-DESUL_INLINE_FUNCTION T
-atomic_oper_fetch(const Oper& op,
-                  T* const dest,
-                  typename std::enable_if<sizeof(T) == 8, const T>::type val,
-                  MemoryOrder order,
-                  MemoryScope scope) {
-  union U {
-    int64_t i;
-    T t;
-    DESUL_INLINE_FUNCTION U() {}
-  } oldval, assume, newval;
-
-  oldval.t = *dest;
+template<>
+struct atomic_compare_exchange_type<4> {
+  using type = int32_t;
+};
 
-  do {
-    assume.i = oldval.i;
-    newval.t = op.apply(assume.t, val);
-    oldval.i = desul::atomic_compare_exchange(
-        (int64_t*)dest, assume.i, newval.i, order, scope);
-  } while (assume.i != oldval.i);
+template<>
+struct atomic_compare_exchange_type<8> {
+  using type = int64_t;
+};
 
-  return newval.t;
-}
+template<>
+struct atomic_compare_exchange_type<16> {
+  using type = Dummy16ByteValue;
+};
 
-#if defined(DESUL_HAVE_16BYTE_COMPARE_AND_SWAP)
+template<class T>
+struct dont_deduce_this_parameter { using type = T; };
+template<class T>
+using dont_deduce_this_parameter_t = typename dont_deduce_this_parameter<T>::type;
 
-template <class Oper, typename T, class MemoryOrder, class MemoryScope>
+template <class Oper, typename T, class MemoryOrder, class MemoryScope,
+  // equivalent to:
+  //   requires atomic_always_lock_free(sizeof(T))
+  std::enable_if_t<atomic_always_lock_free(sizeof(T)), int> = 0
+>
 DESUL_INLINE_FUNCTION T
 atomic_fetch_oper(const Oper& op,
                   T* const dest,
-                  typename std::enable_if<sizeof(T) == 16, const T>::type val,
+                  dont_deduce_this_parameter_t<const T> val,
                   MemoryOrder order,
                   MemoryScope scope) {
-  union U {
-    Dummy16ByteValue i;
-    T t;
-    DESUL_INLINE_FUNCTION U() {}
-  } oldval, assume, newval;
-
-  oldval.t = *dest;
+  using cas_t = typename atomic_compare_exchange_type<sizeof(T)>::type;
+  cas_t oldval = reinterpret_cast<cas_t&>(*dest);
+  cas_t assume = oldval;
 
   do {
-    assume.i = oldval.i;
-    newval.t = op.apply(assume.t, val);
-    oldval.i = desul::atomic_compare_exchange(
-        (Dummy16ByteValue*)dest, assume.i, newval.i, order, scope);
-  } while (assume.i != oldval.i);
+    assume = oldval;
+    T newval = op.apply(reinterpret_cast<T&>(assume), val);
+    oldval = desul::atomic_compare_exchange(
+        reinterpret_cast<cas_t*>(dest), assume, reinterpret_cast<cas_t&>(newval), order, scope);
```

Comment:
In principle you can just do
```suggestion
        reinterpret_cast<cas_t*>(dest), assume, reinterpret_cast<cas_t>(newval), order, scope);
```
since the argument is taken by value.

I think the `reinterpret_cast<T&>(val)` was suggested as alternative to our doing `*reinterpret_cast<T*>(&val)` but here the reference is actually superfluous.

Reply:
its not since its reinterpret_cast and that can only do references or pointers if I am not mistaken ...

Reply:
Your right.  Not sure what I was thinking :)